### PR TITLE
ES|QL Add primitive float variants of all aggregators to the compute engine

### PR DESCRIPTION
**Pull request by ChrisHegarty**
This commit adds primitive float variants of all ES|QL aggregators to the compute engine. 

Much of the changes are mechanical and/or are generated, and retain similar type input/output params as other aggregators.

| Agg   | input type | output type |
|----------|----------|----------|
| Min/Max    | float   | float   |
| Sum    | float   | double   |
| Count Distinct  | float   | long   |
| Median Abs Dev    | float   | double   |
| Percentile  | float   | double   |
| Top List  | float   | float   |
| Values  | float   | float   |

Still no language changes or support for floats at the language level, this PR just addresses support in the engine.

relates: #109746

### CODE REVIEW

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ArrayState.java.st [hunk 1]
```diff
@@ -20,6 +20,9 @@ $endif$
 $if(double)$
 import org.elasticsearch.compute.data.IntVector;
 $endif$
+$if(float)$
```
Comment (costin) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ArrayState.java.st:23]: nit: how about merging it with the others on top?
```
$if(double|float)$
$endif$
```

Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ArrayState.java.st]: Done - [d9c2f55](https://github.com/elastic/elasticsearch/pull/109781/commits/d9c2f556218bd63342df2e8cac322d3b38b67772)

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ForkingOperatorTestCase.java [hunk 2]
```diff
@@ -353,4 +354,26 @@ public void setThreadPool() {
...
+
+        @Override
+        public int compare(Float o1, Float o2) {
+            float first = o1;
+            float second = o2;
+            if (first < second) {
+                return -1;
+            } else if (first == second) {
+                return 0;
+            } else {
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ForkingOperatorTestCase.java:374]: I guess we can assume `o1` and `o2` to never be `NaN`?

**Review by alex-spies — COMMENTED** [review]
Got a first pass; still need to review the tests in a second pass.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ValuesAggregator.java.st [hunk 3]
```diff
@@ -98,6 +100,12 @@ $elseif(int)$
          * the top 32 bits for the group, the bottom 32 for the value.
          */
         state.values.add((((long) groupId) << Integer.SIZE) | (v & 0xFFFFFFFFL));
+$elseif(float)$
+        /*
+         * Encode the groupId and value into a single long -
+         * the top 32 bits for the group, the bottom 32 for the value.
+         */
+        state.values.add((((long) groupId) << Float.SIZE) | (Float.floatToIntBits(v) & 0xFFFFFFFFL));
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ValuesAggregator.java.st:108]: Maybe rounding. I know I just do bit shifting stuff if I want to pack values like this so I don't have to think about the casting rules.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ForkingOperatorTestCase.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ForkingOperatorTestCase.java:374]: We never make NaN or +Inf or -Inf from out functions. Those always become `null`. Might be sane to be paranoid here and assert they aren't those things. Unsure, just drive by comment.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ValuesAggregator.java.st (hunk 3 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ValuesAggregator.java.st:108]: There's also "I'm packing values into a float and it's normal mask the low bits like this when I do that". It's "normal" if you are packing stuff.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CountDistinctFloatAggregator.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.elasticsearch.compute.ann.GroupingAggregator;
+import org.elasticsearch.compute.ann.IntermediateState;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.LongBlock;
+import org.elasticsearch.compute.operator.DriverContext;
+
+@Aggregator({ @IntermediateState(name = "hll", type = "BYTES_REF") })
+@GroupingAggregator
+public class CountDistinctFloatAggregator {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CountDistinctFloatAggregator.java:22]: We call `static` methods in generated code so abstract superclass may be annoying to handle. OTOH, we can generated this with `stringtemplate` and call it a day. I've taken to doing that for some of these things that are just the same for every type except signature. It's like having reified generics the hard way.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/MinFloatAggregator.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import org.elasticsearch.compute.ann.Aggregator;
+import org.elasticsearch.compute.ann.GroupingAggregator;
+import org.elasticsearch.compute.ann.IntermediateState;
+
+@Aggregator({ @IntermediateState(name = "min", type = "FLOAT"), @IntermediateState(name = "seen", type = "BOOLEAN") })
+@GroupingAggregator
+class MinFloatAggregator {
+
+    public static float init() {
+        return Float.POSITIVE_INFINITY;
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/MinFloatAggregator.java:19]: I haven't checked, but could we accidentally emit this or are these sentinals for `null` in the results?

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/Types.java [hunk 6]
```diff
@@ -234,6 +251,12 @@ static ClassName builderType(TypeName resultType) {
         if (resultType.equals(DOUBLE_VECTOR)) {
             return DOUBLE_VECTOR_BUILDER;
         }
+        if (resultType.equals(FLOAT_BLOCK)) {
+            return FLOAT_BLOCK_BUILDER;
+        }
+        if (resultType.equals(FLOAT_VECTOR)) {
+            return DOUBLE_VECTOR_BUILDER;
```
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/Types.java:258]: oops! good catch.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/MinFloatAggregator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/MinFloatAggregator.java:19]: I just checked, and at least when run end-to-end, a `stats max(field)` and `stats min(field)` just return `null` when there are no values for `field`. So both positive infinity and max value should be fine in practical terms.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MaxFloatAggregatorFunctionTests.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    protected String expectedDescriptionOfAggregator() {
+        return "max of floats";
+    }
+
+    @Override
+    public void assertSimpleOutput(List<Block> input, Block result) {
+        Float max = input.stream().flatMap(AggregatorFunctionTestCase::allFloats).map(Float.class::cast).max(floatComparator()).get();
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MaxFloatAggregatorFunctionTests.java:40]: Isn't the cast to `Float` redundant?

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ForkingOperatorTestCase.java (hunk 2 above)
Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ForkingOperatorTestCase.java:374]: Having seen where this is used in the tests, this is 100% fine. Never mind.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MedianAbsoluteDeviationFloatGroupingAggregatorFunctionTests.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        // The input data is small enough that tdigest will find the actual median.
+        Float[] data = s.sorted().toArray(Float[]::new);
+        if (data.length == 0) {
+            return 0;
+        }
+        int c = data.length / 2;
+        return data.length % 2 == 0 ? (data[c - 1] + data[c]) / 2 : data[c];
+    }
+
+    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/101569")
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MedianAbsoluteDeviationFloatGroupingAggregatorFunctionTests.java:80]: The referenced issue is closed; there _should_ not be any reason to mute this; I think we just forgot to unmute.

Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MedianAbsoluteDeviationFloatGroupingAggregatorFunctionTests.java:80]: C.f. https://github.com/elastic/elasticsearch/pull/109942

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        assertThat(((DoubleBlock) result).getDouble(0), closeTo(sum, .0001));
+    }
+
+    public void testOverflowSucceeds() {
+        DriverContext driverContext = driverContext();
+        List<Page> results = new ArrayList<>();
+        try (
+            Driver d = new Driver(
+                driverContext,
+                new SequenceFloatBlockSourceOperator(driverContext.blockFactory(), Stream.of(Float.MAX_VALUE - 1, 2f)),
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java:57]: I don't think this is testing what it should be testing.

The corresponding test for doubles checks what happens when the state overflows. This just checks that we can sum beyond max float, which is still well within limits for the aggregator state (which is a double).

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+                new TestResultPageSinkOperator(results::add),
+                () -> {}
+            )
+        ) {
+            runDriver(d);
+        }
+        assertEquals(sum, results.get(0).<DoubleBlock>getBlock(0).getDouble(0), 1e-10);
+        assertDriverContext(driverContext);
+
+        // Summing up some big double values and expect infinity result
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java:115]: Similarly here; at least, the comment doesn't match the assertion anymore.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+            )
+        ) {
+            runDriver(d);
+        }
+        assertEquals((double) Float.MAX_VALUE * n, results.get(0).<DoubleBlock>getBlock(0).getDouble(0), 0d);
+        assertDriverContext(driverContext);
+
+        results.clear();
+        for (int i = 0; i < n; i++) {
+            largeValues[i] = -Float.MAX_VALUE;
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java:139]: And here.

**Review by alex-spies — APPROVED** [review]
Looks good except for a couple minor comments on the tests; before merging, I think we should probably remove the copy-pasted mute, and we should check if the sum aggregation tests test what they ought to, given that the aggregation state is double not float.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CountDistinctFloatAggregator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CountDistinctFloatAggregator.java:22]: I agree that there is room for improvement here - to generate some of these with templating. Let's separate that out, as it relate to code beyond float.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/SumFloatAggregator.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import org.elasticsearch.compute.ann.IntermediateState;
+
+@Aggregator(
+    {
+        @IntermediateState(name = "value", type = "DOUBLE"),
+        @IntermediateState(name = "delta", type = "DOUBLE"),
+        @IntermediateState(name = "seen", type = "BOOLEAN") }
+)
+@GroupingAggregator
+class SumFloatAggregator extends SumDoubleAggregator {
```
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/SumFloatAggregator.java:21]: Inheritance avoids a lot of unnecessary duplication.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MaxFloatAggregatorFunctionTests.java (hunk 7 above)
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/MaxFloatAggregatorFunctionTests.java:40]: good catch.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java (hunk 9 above)
Comment (ChrisHegarty) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/SumFloatAggregatorFunctionTests.java:57]: Right. It's just testing that we handle on the boundary of Float values, not double here.

**Review by ChrisHegarty — COMMENTED** [review]
Thanks for committing the TopList float aggregator @ivancea

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CountDistinctFloatAggregator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CountDistinctFloatAggregator.java:22]: :+1:

**Review by nik9000 — APPROVED** [review]

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Pull request merged**